posthog/api/kafka_inspector.py

```diff
@@ -0,0 +1,85 @@
+from typing import Union
+
+from kafka import TopicPartition
+from rest_framework import serializers, viewsets
+from rest_framework.decorators import action
+from rest_framework.response import Response
+
+from ee.kafka_client.client import build_kafka_consumer
+from posthog.permissions import IsStaffUser
+
+KAFKA_CONSUMER_TIMEOUT = 1000
+
+# the kafka package doesn't expose ConsumerRecord
+class KafkaConsumerRecord:
+    topic: str
+    partition: int
+    offset: int
+    timestamp: int
+    key: str
+    value: Union[dict, str]
+
+    def __init__(self, topic, partition, offset, timestamp, key, value):
+        self.topic = topic
+        self.partition = partition
+        self.offset = offset
+        self.value = value
+        self.timestamp = timestamp
+        self.key = key
+
+
+class KafkaMessageSerializer(serializers.Serializer):
+    topic = serializers.CharField(read_only=True)
+    partition = serializers.IntegerField(read_only=True)
+    offset = serializers.IntegerField(read_only=True)
+    timestamp = serializers.IntegerField(read_only=True)
+    key = serializers.CharField(read_only=True)
+    value = serializers.JSONField(read_only=True)
+
+
+class KafkaInspectorViewSet(viewsets.ViewSet):
+    permission_classes = [IsStaffUser]
+
+    @action(methods=["POST"], detail=False)
+    def fetch_message(self, request):
+        topic = request.data.get("topic", None)
+        partition = request.data.get("partition", None)
+        offset = request.data.get("offset", None)
+
+        if type(topic) != str:
+            return Response({"error": "Invalid topic."}, status=400)
+
+        if type(partition) != int:
+            return Response({"error": "Invalid partition."}, status=400)
+
+        if type(offset) != int:
+            return Response({"error": "Invalid offset."}, status=400)
+
+        try:
+            message = get_kafka_message(topic, partition, offset)
+            serializer = KafkaMessageSerializer(message, context={"request": request})
+            return Response(serializer.data)
+        except AssertionError:
+            return Response({"error": "Invalid partition/offset pair."}, status=400)
+        except StopIteration:
+            return Response(
+                {
+                    "error": f"Error reading message, most likely the consumer timed out after {KAFKA_CONSUMER_TIMEOUT}ms."
+                },
+                status=400,
+            )
+        except Exception as e:
+            return Response({"error": e.__str__()}, status=500)
+
+
+def get_kafka_message(topic: str, partition: int, offset: int) -> KafkaConsumerRecord:
+    consumer = build_kafka_consumer(
```

Comment:
just to make sure I understand this - we create a new one-off consumer every time. Each kafka consumer group (by group id) tracks their offsets separately, so it doesn't impact any other consumers we already have reading these topics?
What was the problem with we shouldn't do select over CH kafka tables (is that because select applies the one kafka consumer group that's there reading those tables)?

Reply:
Correct.

When a Kafka table is set up, we specify a consumer group ID. Hence any select moves that offset forward.

In this case we use a different consumer group ID, so one will never affect the other. `seek` also doesn't commit offsets, it just moves the memory reference of that offset, but nothing gets stored on the broker end.